[WIP] Github Actions for testing

darcymason opened the pull request:
Early attempts at switching to github actions.  Starting with minimal pytest only, no coverage, etc.

Look away for a while, likely to be ugly for some time...

darcymason commented:
Summary up to 2e9a997:

- no caching yet
- only ~2.5 real-world minutes to run {ubuntu, windows} x {Python 3.6, 3.7, 3.8) x {pytest, install numpy and pytest again}
- each pytest is ~10 sec (my laptop is ~30) so pretty fast env, unless something missing (but output seems fine)
- no working pep8 check yet

pep8speaks commented:
Hello @darcymason! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pydicom/sr/codedict.py`](https://github.com/pydicom/pydicom/blob/3305593af59edf1ac8176f42f62d112a1e07dc0c/pydicom/sr/codedict.py):

> [Line 76:80](https://github.com/pydicom/pydicom/blob/3305593af59edf1ac8176f42f62d112a1e07dc0c/pydicom/sr/codedict.py#L76): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)

darcymason commented:
Some thoughts on next steps/caveats:

* built gdcm here because [recipe for non-conda installing](https://github.com/pydicom/pydicom/wiki/Installing-the-Python-GDCM-bindings-without-Conda) is ubuntu 19.04+ and github actions currently still on 18.04. Also, if cache is fine, then will rarely be built.  However, build is ~5 min.
* need to check cache for gdcm working properly, and is shared across versions etc.
* cache could also be inherited from the master branch so not rebuilt on each separate PR
* using repository release branch of gdcm, so can change on us if a new release has been put out (after cache invalidates by time - no version number here in the key)
* can cache other parts - e.g. [pip installed items](https://github.com/actions/cache/blob/master/examples.md#python---pip)
* currently am not testing full combination of Python version x OS x libraries.  I think it is okay to test many in just one combination, especially on PR commits.  Could test all on merge, perhaps, with a merge workflow.  But with full caching, it may be very quick to do all anyway.
* there may be some combination library tests not covered here: not pillow but gdcm, etc.
* ideally can [mark tests](https://docs.pytest.org/en/latest/example/markers.html) to run just the relevant ones for gdcm, pillow, etc.
* not sure if swig and python3-dev need to be installed again on test run or if cached gdcm is okay without them
* still have to add pytest-cov, generate artifacts, and see if can combine them, or fall back to another hook for handling coverage
* still to add in Simon's pymedphys-pydicom tests, but perhaps doing it on merge workflow is sufficient

scaramallion commented:
This is what I've been using for coverage on the pynetdicom actions testing:
```
- name: Send coverage results
  if: ${{ success() }}
  run: |
    bash <(curl --connect-timeout 10 --retry 10 --retry-max-time \
    0 https://codecov.io/bash) || (sleep 30 && bash <(curl \
    --connect-timeout 10 --retry 10 --retry-max-time \
    0 https://codecov.io/bash))
```

scaramallion commented:
It might also make sense to turn off [fail-fast](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast), to help find other failures that may occur?

darcymason commented:
> This is what I've been using for coverage on the pynetdicom actions testing

Thanks for that -- I'll look into it.

> It might also make sense to turn off [fail-fast](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast) to help find other failures that may occur

I think I like having that on just to save time, with the assumption that the first failing pytest run will normally catch all the errors anyway unless actually working on pixel handlers, or for some weird thing specific to Windows.  On the other hand, if we can get the times down to a few minutes for all this then turning it off would be fine.

SimonBiggs commented:
By the way, the following is the only line needed to make it so that a certain job will only run if another one has completed and passed:

https://github.com/pymedphys/pymedphys/blob/863baac19626408da0a0eabd86e652c19a0fef68/.github/workflows/on-push.yml#L330

darcymason commented:
So ... some input appreciated.

Here is where we are up to fc6fe65 :

|   OS    | Python |       flake8       | pytest-cov | no deps | +numpy | +Pillow and CharPyLs | +GDCM
| :-----: | :----: | :----------------: | :---: |  :----:  |  :----:   |  :----: | :---:
Ubuntu  |  3.6   | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | *  | *
||  3.7   || :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | * | :heavy_check_mark:
||  3.8   || :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | [1]
Windows |  3.6   | | * | :heavy_check_mark: | :heavy_check_mark: | * | *
 |        |  3.7   || * | :heavy_check_mark: | :heavy_check_mark: | *| *
 |        |  3.8   || * | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:| *

`*` = propose for master workflow
[1] = I thought I saw somewhere there were problems on Python 3.8?  Can't find it now.  In any case, the build is not / should not be Python dependent.

Notes:
* Run time is < ~5 min with cache and max 4 parallel
* Note the "+" in the test headers.  e.g. GDCM is added on top of Pillow/CharPyLs.  No current env with only numpy + GDCM.  -- ❔ 
* Coverage is currently running for Windows, and not uploaded.  But for push workflow I propose to run the Windows tests without --cov.  Really could run on just one Python too, OR, run one Python on both {ununtu, windows}. AFAIK, we don't have any OS-specific code, or Python version-specific code  - ❔
* No macos or PyPy  here, could add those to the master workflow
* flake8 currently running on all files, and first for serious errors (to fail early), and then for all, but errors converted to warning only.  However, no PR annotation for this, although there are github actions to post them if we want.

Finishing:
* try to do flake8 just for modified files.  Also drop back to just the serious syntax errors (with fail-fast), and keep PEP8 speaks for the rest?  Alternate is to push annotation to PR ourselves  - ❔

* drop appveyor, travis webhooks
* put circle back in for sphinx builds for now, address that in separate issue
* [separate PR] make a master workflow - could run on PR close, or on a schedule, or ... - ❔

* [separate PR] add @SimonBiggs's pymedphys test run to the master workflow

I've put some extra large ❔ question marks where I'd particularly appreciate your thoughts, but all is open ...

Would like to get this in place fairly soon, once working well enough to use for a while, then refine it through other issues / PRs.  I need to work on other things for the next while...

SimonBiggs commented:
I really like that workflow file. May I borrow some pieces?

On Wed., 3 Jun. 2020, 5:09 am Darcy Mason, <notifications@github.com> wrote:

> So ... some input appreciated.
>
> Here is where we are up to fc6fe65
> <https://github.com/pydicom/pydicom/commit/fc6fe65027ec23435690629b1fe77df04a85adf2>
> :
> OS Python flake8 pytest-cov no deps +numpy +Pillow and CharPyLs +GDCM
> Ubuntu 3.6 ✔️ ✔️ ✔️ ✔️ * *
> 3.7 ✔️ ✔️ ✔️ * ✔️
> 3.8 ✔️ ✔️ ✔️ ✔️ [1]
> Windows 3.6 * ✔️ ✔️ * *
> 3.7 * ✔️ ✔️ * *
> 3.8 * ✔️ ✔️ ✔️ *
>
> * = propose for master workflow
> [1] = I thought I saw somewhere there were problems on Python 3.8? Can't
> find it now. In any case, the build is not / should not be Python dependent.
>
> Notes:
>
>    - Run time is < ~5 min with cache and max 4 parallel
>    - Note the "+" in the test headers. e.g. GDCM is added on top of
>    Pillow/CharPyLs. No current env with only numpy + GDCM. -- ❔
>    - Coverage is currently running for Windows, and not uploaded. But for
>    push workflow I propose to run the Windows tests without --cov. Really
>    could run on just one Python too, OR, run one Python on both {ununtu,
>    windows}. AFAIK, we don't have any OS-specific code, or Python
>    version-specific code - ❔
>    - No macos or PyPy here, could add those to the master workflow
>    - flake8 currently running on all files, and first for serious errors
>    (to fail early), and then for all, but errors converted to warning only.
>    However, no PR annotation for this, although there are github actions to
>    post them if we want.
>
> Finishing:
>
>    -
>
>    try to do flake8 just for modified files. Also drop back to just the
>    serious syntax errors (with fail-fast), and keep PEP8 speaks for the rest?
>    Alternate is to push annotation to PR ourselves - ❔
>    -
>
>    drop appveyor, travis webhooks
>    -
>
>    put circle back in for sphinx builds for now, address that in separate
>    issue
>    -
>
>    [separate PR] make a master workflow - could run on PR close, or on a
>    schedule, or ... - ❔
>    -
>
>    [separate PR] add @SimonBiggs <https://github.com/SimonBiggs>'s
>    pymedphys test run to the master workflow
>
> I've put some extra large ❔ question marks where I'd particularly
> appreciate your thoughts, but all is open ...
>
> Would like to get this in place fairly soon, once working well enough to
> use for a while, then refine it through other issues / PRs. I need to work
> on other things for the next while...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pydicom/pydicom/pull/1116#issuecomment-637749972>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABSBK64CTFCZPZOCBNDJQI3RUVE7RANCNFSM4NPJDSTA>
> .
>

darcymason commented:
> I really like that workflow file. May I borrow some pieces?

Borrow away ... :wink:

darcymason commented:
Just to keep a list going, one more for "Finishing" section:

* [separate PR] Mark tests and update workflow so that only (re-)run the relevant ones for the dependencies, particularly for Pillow, jpeg_ls, and gdcm, perhaps for numpy.

scaramallion commented on .github/workflows/pythonpackage.yml:
```diff
@@ -0,0 +1,92 @@
+        python-version:  [3.6, 3.7, 3.8]
+
+    steps:
+    - uses: actions/checkout@v1
```
I'm a bit curious about this. Why v1?

darcymason replied:
Just accidental - probably from whatever code I 'borrowed' for that part.  I'll change to v2.

scaramallion commented on .github/workflows/pythonpackage.yml:
```diff
@@ -0,0 +1,92 @@
+      run: |
+        python -m pip install wheel 
+        python -m pip install pytest pytest-cov
+    - name: Flake8 checks
```
Would it make more sense to run the flake8 checks as a separate workflow? I'd imagine failing a build early because of a style issue would get annoying very quickly. Except at the end, I'm usually more interested in the unit tests.

scaramallion commented:
* There's no conda package for GDCM + Python 3.8, maybe that's what you're thinking of?
* I agree with only uploading the Ubuntu coverage reports
* Would it make sense to run flake8 checks as a separate workflow? That way we could still get the unit test results and see any style failures

darcymason commented:
> * There's no conda package for GDCM + Python 3.8, maybe that's what you're thinking of?

That could be it.

> * Would it make sense to run flake8 checks as a separate workflow? That way we could still get the unit test results and see any style failures without having to dig into the results manually

Yeah, I thought about that.  I was keeping everything together while developing just for simplicity.  And I liked the fail-fast for serious syntax errors, but that's very unlikely to appear very often (we're all good reliable coders ;-) As I mentioned, we could just drop flake8 _warnings_ in github actions and just let PEP8-speaks comment like it does now.  Or, there are some actions in the marketplace for flake8 that can also insert PEP8-speaks type annotations into the PR conversation.  For simplicity right now, I'd suggest to try to set up a separate workflow, serious errors only, and keep PEP8 speaks.  It would be nice to make it run first and not build the rest if failed -- I'm not sure how to communicate across different workflow files like that, but I can look it up.

darcymason commented:
> I'm not sure how to communicate across different workflow files like that, but I can look it up.

Well that's easy:

"...you can define dependencies on other jobs using the [jobs.<job_id>.needs](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobs) keyword."

mrbean-bremen commented:
Fantastic job!

> Run time is < ~5 min with cache and max 4 parallel

That sounds good! Was the 4 parallel a restriction by GH actions? (don't remember)

> Note the "+" in the test headers. e.g. GDCM is added on top of Pillow/CharPyLs. No current env with only numpy + GDCM

I think that is sufficient, as I wouldn't expect any side effects of an installed Pillow on GDCM.

> Really could run on just one Python too, OR, run one Python on both {ununtu, windows}.

Probably one Python under Ubuntu should suffice. There is only one minor check for the Python version (for Path like objects that are not supported in 3.5), but that is not worth the effort.

> Would it make sense to run flake8 checks as a separate workflow? 

I second this. I'm also for generally replacing Pep8speaks with that flake8 check - at least for me it is more convenient to have these checks together, and also to be able to see them in my own branch build before I make a PR. This way I can easier amend my commits on pep8 errors instead of adding these annoying "Fixed another pep8 warning" commits.

I will do a review tonight, but at a cursory glance this all looks quite good!

mrbean-bremen commented on .github/workflows/pythonpackage.yml:
```diff
@@ -0,0 +1,113 @@
+    needs: flake8
+    runs-on: ${{ matrix.os }}
+    strategy:
+      max-parallel: 4
```
I think the default would be better (e.g. larger). We could easily have all (currently 6) build jobs build in parallel. Maybe just put 8 here...

darcymason commented on .github/workflows/pythonpackage.yml:
```diff
@@ -0,0 +1,113 @@
+      max-parallel: 4
+      matrix:
+        #  ** Update flake8 literals if change os or python-versions
+        os: [ubuntu-latest, windows-latest] # , macos-latest]
```
I did temporarily during testing.  I had planned to put it back for the 'merge' workflow only, since really it is basically linux, so little chance of pydicom code causing an error in mac and not in linux.

I was thinking of PyPy too (they have that as a stock OS) but I'm not sure what we can set up with that -- probably not gdcm but maybe the others can run.  I'm not very familiar with setting it up and running it though.  And I kind of feel that with Python typing, maybe some other speed-up solutions will be appearing.  In any case, will probably leave for its own PR.

darcymason commented on .github/workflows/pythonpackage.yml:
```diff
@@ -0,0 +1,113 @@
+      run: |
+        python -m pip install pillow cython
+        python -m pip install git+https://github.com/Who8MyLunch/CharPyLS
+        pytest --cov=pydicom --cov-append
```
Sure, can do.

darcymason commented:
> That sounds good! Was the 4 parallel a restriction by GH actions? (don't remember)

It was used in their examples - I've deleted the line now and I think the default was up to 6 depending on how busy their servers are.

> Probably one Python under Ubuntu should suffice. There is only one minor check for the Python version (for Path like objects that are not supported in 3.5)

I've changed coverage to one combination - and we are dropping 3.5 now so should probably remove that code.

> You may probably disable pydicom in Travis and AppVeyor before merging this to avoid the build errors.

Yes, for sure.  I was holding off in case other PRs came in while this was being developed.

SimonBiggs commented:
> I've changed coverage to one combination - and we are dropping 3.5 now so should probably remove that code.

Ooo, I didn't know 3.5 was being dropped. I can now make my Zenodo pull request a little less messy :).

SimonBiggs commented on setup.py:
```diff
@@ -35,7 +35,6 @@
     "Intended Audience :: Science/Research",
     "Development Status :: 5 - Production/Stable",
     "Programming Language :: Python",
-    "Programming Language :: Python :: 3.5",
```
Make sure to change the following line also:

https://github.com/pydicom/pydicom/blob/ba939ef2007c91d2b0dc741fdb0d3e81f066b3c5/setup.py#L81

darcymason replied:
> Make sure to change the following line also:
> 
> https://github.com/pydicom/pydicom/blob/ba939ef2007c91d2b0dc741fdb0d3e81f066b3c5/setup.py#L81

Right - I knew that, they it got lost in doing something else...  Will do.

darcymason commented:
Okay, I think this is close enough to merge... one cleanup issue:

* I've attempted to identify files added/modified in the push (hopefully relative to reference branch for the PR) but that does not seem to be working - flake8 is still operating on all files despite attempts to list the specific ones changed.  But since just listing warnings, can live with for now.

Otherwise I'm quite happy with it.  Have changed things from the table in my earlier comment - all special checks/coverage in ubuntu/python 3.8 only.  Have simplified to make it easier to expand to a more complete matrix for a different (more comprehensive) workflow.

I'll merge in the morning in case anyone wants to have another look.

Then we'll see how it works with regular PRs...  😨 More tweaks to come, I'm sure...

mrbean-bremen commented on pydicom/fileutil.py:
```diff
@@ -283,10 +283,6 @@ def path_from_pathlike(file_object):
         ``PathLike`` objects have been introduced in Python 3.6. In Python 3.5,
         only objects of type :class:`pathlib.Path` are considered.
```
May want to remove the 3.5 part here. Also I think I wrote a 3.5-specific test, or at least excluded 3.5 from some tests.
